### PR TITLE
[ckpt, ci] fix: refactor merge_dcp_to_hf.py to remove internal dependencies and add checkpoint verification

### PR DESCRIPTION
**Pull request by TimYangst**
### What does this PR do?

> Refactors the DCP-to-HuggingFace conversion script to be self-contained and adds checkpoint verification utilities to ensure conversion correctness.

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: https://github.com/ByteDance-Seed/VeOmni/pulls?q=is%3Apr+checkpoint+conversion
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}`: `ckpt, test`
  - `{type}`: `fix`
  - No API breaking changes

### Test

> CI tests validate the changes through existing checkpoint save/load tests (`test_trainer_saveload_ep8`, `test_trainer_saveload_ep4`, `test_trainer_saveload_no_ep`). Each test now verifies:
> - DCP checkpoint is created during training
> - HuggingFace conversion completes successfully
> - Converted checkpoint weights match original DCP weights (tensor-by-tensor comparison)

### API and Usage Example

**Command-line usage (unchanged):**
```bash
python scripts/merge_dcp_to_hf.py \
  --load-dir ./checkpoint/global_step_5 \
  --save-dir ./hf_ckpt \
  --model-assets-dir /path/to/model  # optional
```

**No breaking changes** - maintains backward compatibility.

### Design & Code Changes

**1. `scripts/merge_dcp_to_hf.py` - Self-contained script:**
- Removed dependencies on `veomni.checkpoint` and `veomni.models` modules
- Reimplemented using PyTorch DCP APIs directly (`torch.distributed.checkpoint`)
- Processes checkpoints shard-by-shard (each shard ≤2GB, configurable via `--shard-size`)
- Added model-key filtering to exclude optimizer states and training artifacts
- Improved logging and comprehensive docstrings

**2. `tests/checkpoints/checkpoint_verification_utils.py` - New verification utilities:**
- `load_dcp_checkpoint()`: Loads DCP checkpoints for verification
- `load_hf_checkpoint()`: Loads HuggingFace checkpoints (safetensors/pytorch)
- `verify_checkpoint_conversion()`: Tensor-by-tensor comparison with detailed mismatch reporting
- Supports both sharded and non-sharded checkpoints

**3. `tests/checkpoints/test_trainer_saveload.py` - Enhanced tests:**
- All tests now run `merge_dcp_to_hf.py` as subprocess after training
- Added verification step using `checkpoint_verification_utils`
- Ensures converted weights exactly match original DCP checkpoint

**Memory Efficiency:**
The implementation processes checkpoints shard-by-shard to minimize memory usage:
- Each shard (default 2GB) is loaded, converted, saved, then immediately released
- Within each shard, all tensors are loaded at once due to DCP API requirements (`torch.distributed.checkpoint.load` needs pre-allocated tensors)
- Peak memory = max(shard_size) + overhead, not total checkpoint size
- Aggressive garbage collection (`gc.collect()`) after each shard

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md?plain=1#L22): `make commit && make style && make quality`
- [x] Add / Update [the documentation](https://github.com/ByteDance-Seed/VeOmni/blob/main/docs).
- [x] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/ByteDance-Seed/VeOmni/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: Existing end-to-end tests (`test_trainer_saveload_*`) have been updated to cover the refactored code.

### CODE REVIEW

File: veomni/checkpoint/dcp_checkpointer.py [hunk 1]
```diff
@@ -405,10 +409,13 @@ def dcp_to_torch_state_dict(save_checkpoint_path: Union[str, os.PathLike]) -> ST
...
     state_dict: STATE_DICT_TYPE = {}
 
+    # Filter to only load model weights if requested
+    keys = ["model"] if model_only else None
+
     _load_state_dict(
         state_dict,
         storage_reader=FileSystemReader(save_checkpoint_path),
-        planner=_EmptyStateDictLoadPlanner(),
+        planner=_EmptyStateDictLoadPlanner(keys=keys),
```
Comment (piyifan123) [inline veomni/checkpoint/dcp_checkpointer.py]: based on the user reporting, just set keys=['models'] can work?

Comment (TimYangst) [reply veomni/checkpoint/dcp_checkpointer.py]: I added a flag there:  ```keys = ["model"] if model_only else None ```

File: veomni/checkpoint/dcp_checkpointer.py [hunk 2]
```diff
@@ -405,10 +409,13 @@ def dcp_to_torch_state_dict(save_checkpoint_path: Union[str, os.PathLike]) -> ST
     # Load the state_dict from the DCP checkpoint
     state_dict: STATE_DICT_TYPE = {}
 
+    # Filter to only load model weights if requested
+    keys = ["model"] if model_only else None
```
Comment (piyifan123) [inline veomni/checkpoint/dcp_checkpointer.py]: so none means loading everything? please comment to clarify.

File: veomni/models/module_utils.py [hunk 3]
```diff
@@ -119,37 +122,45 @@ class BroadcastMetadata:
     dtype: Optional["torch.dtype"]
 
 
-def _load_state_dict(weights_path: str, **kwargs) -> List["StateDictIterator"]:
+def _load_state_dict(
```
Comment (piyifan123) [inline veomni/models/module_utils.py]: do not quite understand why calling this works.. from the comment and the logic, it loads a ckpt in transformers (i.e., HF) format, however, the script merge_dcp_to_hf.py loads a dcp format ckpt to state dict? 

anyway, is it possible to write a new merge_dcp_to_hf.py to make it self-contained and add a step in https://github.com/ByteDance-Seed/VeOmni/blob/main/tests/checkpoints/test_trainer_saveload.py to test?

File: veomni/checkpoint/dcp_checkpointer.py (hunk 2 above)
Comment (TimYangst) [reply veomni/checkpoint/dcp_checkpointer.py]: Done

**Review by Luosuu — CHANGES_REQUESTED** [review]
I think we can add test for this method in our CI. Maybe just in the `tests/checkpoints/test_trainer_saveload.py`?

**Comment by Coach257** [comment]
rebase下master可以过ci的autolabel

**Comment by FoolPlayer** [comment]
> I think we can add test for this method in our CI. Maybe just in the `tests/checkpoints/test_trainer_saveload.py`?

agree

File: veomni/models/module_utils.py (hunk 3 above)
Comment (TimYangst) [reply veomni/models/module_utils.py]: Done, rewritten the script and added the test.

**Comment by TimYangst** [comment]
> I think we can add test for this method in our CI. Maybe just in the `tests/checkpoints/test_trainer_saveload.py`?

Done

File: tests/checkpoints/test_trainer_saveload.py [hunk 4]
```diff
@@ -344,43 +353,139 @@ def step_id(s):
...
+
+    # Verify the output exists
+    assert os.path.exists(hf_output_dir), f"HF checkpoint directory not found: {hf_output_dir}"
+    files = os.listdir(hf_output_dir)
+    logger.info(f"HF checkpoint files: {files}")
+    assert len(files) > 0, "No files generated in HF checkpoint directory"
+
+    # Verify HF checkpoint structure
+    logger.info("Verifying HF checkpoint structure...")
+    assert verify_hf_checkpoint(hf_output_dir, verify_structure=True, verify_weights=False), (
```
Comment (piyifan123) [inline tests/checkpoints/test_trainer_saveload.py]: why is verify_weight is false

File: tests/checkpoints/test_trainer_saveload.py [hunk 5]
```diff
@@ -344,43 +353,139 @@ def step_id(s):
...
```
Comment (piyifan123) [inline tests/checkpoints/test_trainer_saveload.py]: same here.

File: scripts/merge_dcp_to_hf.py [hunk 6]
```diff
@@ -1,40 +1,349 @@
...
+
+    # Only load model weights (excluding optimizer states, etc.)
+    # keys=["model"] will only load model state dict
+    # See _EmptyStateDictLoadPlanner._should_include_key() in torch's default_planner.py
+    keys = ["model"]
+
+    _load_state_dict(
+        state_dict,
+        storage_reader=FileSystemReader(save_checkpoint_path),
+        planner=_EmptyStateDictLoadPlanner(keys=keys),
```
Comment (piyifan123) [inline scripts/merge_dcp_to_hf.py]: so it still loads the entire ckpt to the cpu mem? just asking, it's OK to have that at the beginning.

File: scripts/merge_dcp_to_hf.py [hunk 7]
```diff
@@ -1,40 +1,349 @@
...
+    This function only loads model weights (ignoring optimizer states, etc.) to reduce memory usage.
+
+    Args:
+        save_checkpoint_path: Directory containing the DCP checkpoint.
+
+    Returns:
+        state_dict: Dictionary containing model weights.
+
+    Warning:
+        To avoid OOM, it's recommended to only run this function on a single rank.
```
Comment (piyifan123) [inline scripts/merge_dcp_to_hf.py]: what does this mean? we should only run this on a CPU machine. And even if GPU is available, we should avoid use GPU mem to load dcp as it's only 80GB most times which is insufficient for bigger models.

File: scripts/merge_dcp_to_hf.py [hunk 8]
```diff
@@ -1,40 +1,349 @@
...
+
+    # Load the state_dict from the DCP checkpoint
+    state_dict: STATE_DICT_TYPE = {}
+
+    # Only load model weights (excluding optimizer states, etc.)
+    # keys=["model"] will only load model state dict
+    # See _EmptyStateDictLoadPlanner._should_include_key() in torch's default_planner.py
+    keys = ["model"]
+
+    _load_state_dict(
```
Comment (piyifan123) [inline scripts/merge_dcp_to_hf.py]: this is the public API from torch, wondering if we can use this https://github.com/pytorch/pytorch/blob/orig/release/2.8/torch/distributed/checkpoint/state_dict_loader.py#L53-L61

Comment (TimYangst) [reply scripts/merge_dcp_to_hf.py]: done

File: tests/checkpoints/test_trainer_saveload.py (hunk 4 above)
Comment (TimYangst) [reply tests/checkpoints/test_trainer_saveload.py]: Updated.

File: tests/checkpoints/test_trainer_saveload.py (hunk 5 above)
Comment (TimYangst) [reply tests/checkpoints/test_trainer_saveload.py]: Updated

File: scripts/merge_dcp_to_hf.py (hunk 7 above)
Comment (TimYangst) [reply scripts/merge_dcp_to_hf.py]: Tested on cpu only machine, it should work

File: scripts/merge_dcp_to_hf.py (hunk 6 above)
Comment (TimYangst) [reply scripts/merge_dcp_to_hf.py]: Yes, you're right. The current implementation does load the entire checkpoint into CPU memory at once. This is required by the `torch.distributed.checkpoint.load` API design, which needs pre-allocated tensors with correct shapes to perform in-place loading.

For future optimization, we could explore:
- Incremental loading: Process tensors shard-by-shard instead of loading everything upfront
- Streaming conversion: Load → convert → save in a pipeline to avoid holding two copies in memory
- Memory-mapped reading: Reduce memory footprint for very large checkpoints

But as an initial implementation, this approach prioritizes correctness and simplicity.

File: scripts/merge_dcp_to_hf.py [hunk 9]
```diff
@@ -1,40 +1,323 @@
...
+    all_dcp_keys = []
+
+    for key, tensor_meta in metadata.state_dict_metadata.items():
+        hf_key = _normalize_key(key)
+        if hf_key:
+            # Determine dtype for size calculation
+            if save_dtype:
+                dtype = getattr(torch, save_dtype) if isinstance(save_dtype, str) else save_dtype
+            else:
+                dtype = tensor_meta.properties.dtype if hasattr(tensor_meta.properties, "dtype") else torch.float32
```
Comment (Luosuu) [inline scripts/merge_dcp_to_hf.py]: is it safe to assume dtype to be fp32 when we cannot read its dtype?

File: scripts/merge_dcp_to_hf.py [hunk 10]
```diff
@@ -1,40 +1,323 @@
...
+        torch.float16: 2,
+        torch.bfloat16: 2,
+        torch.int64: 8,
+        torch.int32: 4,
+        torch.int16: 2,
+        torch.int8: 1,
+        torch.uint8: 1,
+        torch.bool: 1,
+    }
+    return size_map.get(dtype, 4)
```
Comment (Luosuu) [inline scripts/merge_dcp_to_hf.py:44]: same here, is default value of 32 safe?

File: scripts/merge_dcp_to_hf.py (hunk 9 above)
Comment (Luosuu) [reply scripts/merge_dcp_to_hf.py]: let's assert they must have dtensor metadata

File: tests/checkpoints/checkpoint_verification_utils.py [hunk 11]
```diff
@@ -0,0 +1,428 @@
...
+        if hasattr(tensor, "full_tensor"):
+            tensor = tensor.full_tensor()
+
+        # Remove the first "model." prefix for HuggingFace format
+        # DCP keys are like "model.model.layers.0.weight"
+        # HF keys should be like "model.layers.0.weight"
+        # Remove the first "model." prefix for HuggingFace format
+        # DCP keys are like "model.model.layers.0.weight"
+        # HF keys should be like "model.layers.0.weight"
+        hf_key = key[6:]  # Remove "model." prefix (6 characters)
```
Comment (piyifan123) [inline tests/checkpoints/checkpoint_verification_utils.py]: if key.startswith(model.model) then truncate.

**Review by FoolPlayer — APPROVED** [review]

**Comment by FoolPlayer** [comment]
LGTM! Can you help add the usage in docs in next PR, thanks!

File: scripts/merge_dcp_to_hf.py (hunk 9 above)
Comment (TimYangst) [reply scripts/merge_dcp_to_hf.py]: Done, raised an exception there. And discussed with yifan last night, we will convert the weight type as bf16 as default.

File: scripts/merge_dcp_to_hf.py (hunk 10 above)
Comment (TimYangst) [reply scripts/merge_dcp_to_hf.py:44]: As we raised an exception below, we won't go to default value there.

File: tests/checkpoints/checkpoint_verification_utils.py (hunk 11 above)
Comment (TimYangst) [reply tests/checkpoints/checkpoint_verification_utils.py]: Done, I add log in the merge_dcp_to_hf script, in case, there other key that only have single model. prefix.

**Comment by TimYangst** [comment]
> LGTM! Can you help add the usage in docs in next PR, thanks!

Sure, will have a follow up MR for the usage doc.

**Review by Luosuu — APPROVED** [review]
